### PR TITLE
Chore: Catch `fetchContent` exception in rules

### LINKED ISSUE #977: Some rules need to catch fetch exception
In some cases, fetch a resource can throw an exception. If we don't capture the exception, the scan will finish without a result.

We need to catch the exception in rules that are not catching the exception.

* [ ] `no-bom`
* [ ] `http-compression`

### PR DESCRIPTION
**Pull request by sarvaje**
Fix #977

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/getting-started/pull-requests/#commit-messagess)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [ ] Added/Updated related tests.

### CODE REVIEW

File: packages/rule-http-compression/src/rule.ts [hunk 1]
```diff
@@ -259,8 +259,17 @@ export default class HttpCompressionRule implements IRule {
...
         const checkBrotli = async (resource, element) => {
-            const { contentEncodingHeaderValue, rawResponse, response } = await getNetworkData(resource, { 'Accept-Encoding': 'br' });
+            let contentEncodingHeaderValue;
+            let rawResponse;
+            let response;
 
+            try {
+                ({ contentEncodingHeaderValue, rawResponse, response } = await getNetworkData(resource, { 'Accept-Encoding': 'br' }));
+            } catch (err) {
+                await context.report(resource, element, 'Error fetching the content');
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: Maybe give more information in what situation the request failed.

e.g.  `Error fetching the content when requesting the content compress with Brotli`

Comment (sarvaje) [reply packages/rule-http-compression/src/rule.ts]: ok!

**Comment by sarvaje** [comment]
> Can you also add some tests? Thanks!

ok!

**Comment by sarvaje** [comment]
@alrra tests added

File: packages/rule-http-compression/src/rule.ts [hunk 2]
```diff
@@ -456,6 +492,16 @@ export default class HttpCompressionRule implements IRule {
                      * not being served with `content-encoding: gzip`.
                      */
 
+                    const safeRawResponse = asyncTry<Buffer>(response.body.rawResponse);
```
Comment (molant) [inline packages/rule-http-compression/src/rule.ts]: Actually I was thinking about changing this at the connector level

File: packages/rule-http-compression/src/rule.ts [hunk 3]
```diff
@@ -259,8 +264,15 @@ export default class HttpCompressionRule implements IRule {
         };
 
         const checkBrotli = async (resource, element) => {
-            const { contentEncodingHeaderValue, rawResponse, response } = await getNetworkData(resource, { 'Accept-Encoding': 'br' });
+            let networkData = await getNetworkData(resource, { 'Accept-Encoding': 'br' });
+
+            if (!networkData) {
+                await context.report(resource, element, 'Error fetching brotli content');
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: Maybe we should change these types of messages to something more user friendly? 

e.g.: `Couldn't fetch content when Brotli compression is requested` ?

File: packages/rule-http-compression/src/rule.ts [hunk 4]
```diff
@@ -318,18 +330,34 @@ export default class HttpCompressionRule implements IRule {
...
             // Check for user agent sniffing.
 
-            const { rawResponse: uaRawResponse } = await getNetworkData(resource, {
+            networkData = await getNetworkData(resource, {
                 'Accept-Encoding': 'br',
                 'User-Agent': uaString
             });
 
+            if (!networkData) {
+                await context.report(resource, element, 'Error fetching brotli content');
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: > Brotli

File: packages/rule-http-compression/src/rule.ts (hunk 2 above)
Comment (alrra) [reply packages/rule-http-compression/src/rule.ts]: @molant Can you provide more details?

Comment (molant) [reply packages/rule-http-compression/src/rule.ts]: The connector should have the try/catch and return null if it fails so we don't have to wrap it in here.

File: packages/rule-no-bom/package.json [hunk 5]
```diff
@@ -56,7 +56,7 @@
     "lint:js": "eslint . --cache --ext js --ext md --ext ts --ignore-path ../../.eslintignore --report-unused-disable-directives",
     "lint:md": "markdownlint --ignore CHANGELOG.md *.md",
     "sonarwhal": "node node_modules/sonarwhal/dist/src/bin/sonarwhal.js",
-    "test": "npm run lint && npm run build && nyc ava",
+    "test": "npm run build && nyc ava",
```
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/rule-no-bom/package.json]: I deleted it for testing in my computer for some reason and I forgot to put it back hehehe

File: packages/rule-http-compression/src/rule.ts [hunk 6]
```diff
@@ -456,6 +492,16 @@ export default class HttpCompressionRule implements IRule {
                      * not being served with `content-encoding: gzip`.
                      */
 
+                    const safeRawResponse = asyncTry<Buffer>(response.body.rawResponse);
+
+                    const rawResponse: Buffer = await safeRawResponse();
+
+                    if (!rawResponse) {
+                        await context.report(resource, element, 'Error fetching rawResponse');
```
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/rule-http-compression/src/rule.ts]: ok!

**Comment by sarvaje** [comment]
@alrra done!

File: packages/rule-http-compression/src/rule.ts [hunk 7]
```diff
@@ -318,18 +336,34 @@ export default class HttpCompressionRule implements IRule {
...
                 await context.report(resource, element, generateCompressionMessage('Brotli', false, `over HTTPS, regardless of the user agent`));
             }
         };
 
         const checkGzipZopfli = async (resource: string, element: IAsyncHTMLElement, shouldCheckIfCompressedWith: CompressionCheckOptions) => {
-            const { contentEncodingHeaderValue, rawContent, rawResponse, response } = await getNetworkData(resource, { 'Accept-Encoding': 'gzip' });
+            let networkData = await getNetworkData(resource, { 'Accept-Encoding': 'gzip' });
+
+            if (!networkData) {
+                await context.report(resource, element, 'Error fetching gzip content');
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: For consistency, let's change this to: 

> Could not be fetched when requested compressed with gzip

File: packages/rule-http-compression/src/rule.ts [hunk 8]
```diff
@@ -377,11 +411,19 @@ export default class HttpCompressionRule implements IRule {
...
             // Check for user agent sniffing.
 
-            const { rawResponse: uaRawResponse } = await getNetworkData(resource, {
+            networkData = await getNetworkData(resource, {
                 'Accept-Encoding': 'gzip',
                 'User-Agent': uaString
             });
 
+            if (!networkData) {
+                await context.report(resource, element, 'Error fetching gzip content');
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: > Could not be fetched when requested compressed with gzip

File: packages/rule-http-compression/src/rule.ts [hunk 9]
```diff
@@ -500,7 +553,15 @@ export default class HttpCompressionRule implements IRule {
              *     the user agent. "
              */
 
-            const { contentEncodingHeaderValue, rawResponse } = await getNetworkData(resource, { 'Accept-Encoding': 'identity' });
+            const networkData = await getNetworkData(resource, { 'Accept-Encoding': 'identity' });
+
+            if (!networkData) {
+                await context.report(resource, element, 'Error fetching uncompressed content');
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: > Could not be fetched when requested uncompressed

File: packages/rule-http-compression/src/rule.ts (hunk 2 above)
Comment (alrra) [reply packages/rule-http-compression/src/rule.ts]: @molant Can you open an issue?

File: packages/rule-http-compression/src/rule.ts [hunk 10]
```diff
@@ -259,8 +270,15 @@ export default class HttpCompressionRule implements IRule {
         };
 
         const checkBrotli = async (resource, element) => {
-            const { contentEncodingHeaderValue, rawResponse, response } = await getNetworkData(resource, { 'Accept-Encoding': 'br' });
+            let networkData = await getNetworkData(resource, { 'Accept-Encoding': 'br' });
+
+            if (!networkData) {
+                await context.report(resource, element, `Couldn't fetch content when Brotli compression is requested`);
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: Let's use:

> Could not be fetched when requested compressed with Brotli

File: packages/rule-http-compression/src/rule.ts [hunk 11]
```diff
@@ -318,18 +336,34 @@ export default class HttpCompressionRule implements IRule {
...
             // Check for user agent sniffing.
 
-            const { rawResponse: uaRawResponse } = await getNetworkData(resource, {
+            networkData = await getNetworkData(resource, {
                 'Accept-Encoding': 'br',
                 'User-Agent': uaString
             });
 
+            if (!networkData) {
+                await context.report(resource, element, `Couldn't fetch content when Brotli compression is requested`);
```
Comment (alrra) [inline packages/rule-http-compression/src/rule.ts]: > Could not be fetched when requested compressed with Brotli

**Pull request closed without merging**